pkg/controllers/logging.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright The Ratify Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"runtime/debug"
+
+	"github.com/go-logr/logr"
+	"github.com/sirupsen/logrus"
+)
+
+// LogrusSink is an adapter to allow the use of logrus with logr, as required by k8s controller-runtime.
+type LogrusSink struct {
+	logger        *logrus.Logger
+	names         []string
+	keysAndValues []interface{}
+}
+
+func NewLogrusSink(logger *logrus.Logger) *LogrusSink {
+	return &LogrusSink{
+		logger: logger,
+	}
+}
+
+// Init receives optional information about the logr library for LogSink
+// implementations that need it.
+func (sink *LogrusSink) Init(info logr.RuntimeInfo) {
+}
+
+// Enabled tests whether this LogSink is enabled at the specified V-level.
+// For example, commandline flags might be used to set the logging
+// verbosity and disable some info logs.
+func (sink *LogrusSink) Enabled(level int) bool {
+	// controller-runtime uses V=0 for Info and V=1 for Debug: https://github.com/kubernetes-sigs/controller-runtime/blob/master/TMP-LOGGING.md
+	switch level {
+	case 0:
+		return sink.logger.IsLevelEnabled(logrus.InfoLevel)
+	default:
+		return sink.logger.IsLevelEnabled(logrus.DebugLevel)
+	}
+}
+
+// Info logs a non-error message with the given key/value pairs as context.
+// The level argument is provided for optional logging.  This method will
+// only be called when Enabled(level) is true. See Logger.Info for more
+// details.
+func (sink *LogrusSink) Info(level int, msg string, keysAndValues ...interface{}) {
+	entry := sink.logger.WithField("name", sink.names)
+
+	if keysAndValues != nil {
+		entry.WithField("values", keysAndValues)
+	}
+
+	entry.Info(msg, keysAndValues)
+}
+
+// Error logs an error, with the given message and key/value pairs as
+// context.  See Logger.Error for more details.
+func (sink *LogrusSink) Error(err error, msg string, keysAndValues ...interface{}) {
```

Comment:
Curious, would we need to support debug too?

Reply:
I don't think so in this case. We're implementing the `logr.LogSink` interface here (https://pkg.go.dev/github.com/go-logr/logr#LogSink)

`logr` doesn't have named levels, it splits out regular logs (`Info`, with an optional `level` which doesn't have a prescribed meaning) and error logs (`Error`), but that seems to be about it

```go
type LogSink interface {
	// Init receives optional information about the logr library for LogSink
	// implementations that need it.
	Init(info [RuntimeInfo](https://pkg.go.dev/github.com/go-logr/logr#RuntimeInfo))

	// Enabled tests whether this LogSink is enabled at the specified V-level.
	// For example, commandline flags might be used to set the logging
	// verbosity and disable some info logs.
	Enabled(level [int](https://pkg.go.dev/builtin#int)) [bool](https://pkg.go.dev/builtin#bool)

	// Info logs a non-error message with the given key/value pairs as context.
	// The level argument is provided for optional logging.  This method will
	// only be called when Enabled(level) is true. See Logger.Info for more
	// details.
	Info(level [int](https://pkg.go.dev/builtin#int), msg [string](https://pkg.go.dev/builtin#string), keysAndValues ...interface{})

	// Error logs an error, with the given message and key/value pairs as
	// context.  See Logger.Error for more details.
	Error(err [error](https://pkg.go.dev/builtin#error), msg [string](https://pkg.go.dev/builtin#string), keysAndValues ...interface{})

	// WithValues returns a new LogSink with additional key/value pairs.  See
	// Logger.WithValues for more details.
	WithValues(keysAndValues ...interface{}) [LogSink](https://pkg.go.dev/github.com/go-logr/logr#LogSink)

	// WithName returns a new LogSink with the specified name appended.  See
	// Logger.WithName for more details.
	WithName(name [string](https://pkg.go.dev/builtin#string)) [LogSink](https://pkg.go.dev/github.com/go-logr/logr#LogSink)
}
```

So there's no dedicated `Debug()` or `Trace()` functions to implement. And there's no `level` to interact with or handle inside `Error(err)`

k8s `controller-runtime` uses level 0 for normal logs, and 1+ as "debug": https://github.com/kubernetes-sigs/controller-runtime/blob/master/TMP-LOGGING.md#getting-started-with-structured-logging

Reply:
thanks for the clarification. makes sense.